### PR TITLE
MCLOUD-6135: Every step should throw only StepException

### PR DESCRIPTION
**Pull request by NadiyaS**
### Description

Every scenario step should throw only StepException.
If any error occurred on deploy phase - failed flag should be set and post-deploy skipped.

### Fixed Issues (if relevant)

1. https://jira.corp.magento.com/browse/MCLOUD-6135

### Manual testing scenarios

1. ssh to cloud environment and set read only permissions to file app/etc/env.php `chmod 444 app/etc/env.php`
2. run redeployment
Expected result:
 - deploy phase is failed and flag .deploy_failed is set. Post-deploy script is skipped.

### Release notes

For user-facing changes, add a meaningful release note. For examples, see [Magento Cloud ECE-tools release notes](https://devdocs.magento.com/cloud/release-notes/ece-release-notes.html).

### Associated documentation updates

Add link to Magento DevDocs PR or Issue, if needed.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] Pull request introduces user-facing changes and includes meaningful updates for any required release notes and documentation changes
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: src/Scenario/Processor.php [hunk 1]
```diff
@@ -64,19 +64,26 @@ public function execute(array $scenarios): void
...
             array_walk($steps, function (StepInterface $step, string $name) {
                 $this->logger->debug('Running step: ' . $name);
 
                 $step->execute();
 
                 $this->logger->debug(sprintf('Step "%s" finished', $name));
             });
-        } catch (StepException $stepException) {
+        } catch (Throwable $stepException) {
             try {
```
Comment (oshmyheliuk) [inline src/Scenario/Processor.php]: Let's move nested try/catch block to a new separate method as this method contains too many try/catch blocks and it becomes difficult to read.
```
        } catch (Throwable $stepException) {
            $this->runActions();
            $this->handleException($stepException);
        }
```

File: src/Scenario/Processor.php [hunk 2]
```diff
@@ -64,19 +64,26 @@ public function execute(array $scenarios): void
...
         try {
             $this->mergedScenarios = $this->merger->merge($scenarios);
             $steps = $this->mergedScenarios['steps'];
-
+        } catch (Throwable $exception) {
+            $this->handleException(
+                $exception,
+                sprintf('Unhandled error: [%d] %s', $exception->getCode(), $exception->getMessage())
+            );
+        }
```
Comment (oshmyheliuk) [inline src/Scenario/Processor.php:76]: Please add a newline to separate try/catch blocks

File: src/Step/SkipStep.php [hunk 3]
```diff
@@ -39,9 +39,14 @@ public function __construct(LoggerInterface $logger, string $stepName)
      * Logs the information about step skipping.
      *
      * @return void
+     * @throws StepException
      */
     public function execute()
     {
-        $this->logger->info(sprintf('Step "%s" was skipped', $this->stepName));
+        try {
```
Comment (oshmyheliuk) [inline src/Step/SkipStep.php]: Why did you add try/catch here? Logger method `info` doesn't throw any exceptions.

File: src/Step/SkipStep.php [hunk 4]
```diff
@@ -39,9 +39,14 @@ public function __construct(LoggerInterface $logger, string $stepName)
      * Logs the information about step skipping.
      *
      * @return void
+     * @throws StepException
```
Comment (oshmyheliuk) [inline src/Step/SkipStep.php]: these two lines can be changed to `{@inheritDoc}`

File: src/Scenario/Processor.php (hunk 1 above)
Comment (NadiyaS) [reply src/Scenario/Processor.php]: done

File: src/Scenario/Processor.php (hunk 2 above)
Comment (NadiyaS) [reply src/Scenario/Processor.php:76]: done

File: src/Step/SkipStep.php (hunk 3 above)
Comment (NadiyaS) [reply src/Step/SkipStep.php]: changes reverted

File: src/Step/SkipStep.php (hunk 4 above)
Comment (NadiyaS) [reply src/Step/SkipStep.php]: done

**Review by oshmyheliuk — DISMISSED** [review]

**Review by BarnyShergold — DISMISSED** [review]

**Review by BaDos — DISMISSED** [review]

**Comment by andriyShevtsov** [comment]
QA approved

File: src/Step/Deploy/SplitDbConnection.php [hunk 5]
```diff
@@ -110,56 +110,60 @@ public function __construct(
...
+            $notAvailableSplitTypes = $this->getMissedSplitTypes(
+                $splitTypes,
+                $dbConfig[DbConfig::KEY_CONNECTION] ?? []
+            );
 
-        if (!empty($notAvailableSplitTypes)) {
-            $this->logger->warning(
+            if (!empty($notAvailableSplitTypes)) {
+                $this->logger->warning(
                 sprintf(
```
Comment (BarnyShergold) [inline src/Step/Deploy/SplitDbConnection.php]: Indent of lines 133/137/138 causing tests to fail - should be indented a further 4 spaces.

**Review by BarnyShergold — CHANGES_REQUESTED** [review]

File: src/Step/Deploy/SplitDbConnection.php (hunk 5 above)
Comment (NadiyaS) [reply src/Step/Deploy/SplitDbConnection.php]: fixed

File: src/Scenario/Processor.php [hunk 6]
```diff
@@ -64,43 +63,55 @@ public function execute(array $scenarios): void
...
-                    $this->logger->debug('Running on fail action: ' . $name);
-
-                    $action->execute();
-
-                    $this->logger->debug(sprintf('On fail action "%s" finished', $name));
-                });
-            } catch (ActionException $actionException) {
-                $this->logger->error($actionException->getMessage());
-            }
+        } catch (Throwable $stepException) {
```
Comment (oshmyheliuk) [inline src/Scenario/Processor.php:86]: Looks like here should be `StepException` instead of `Throwable`, as `Step` can throw only `StepException`

Comment (oshmyheliuk) [reply src/Scenario/Processor.php:86]: you can change this method to something like this with one Try/Catch block
```php
    public function execute(array $scenarios): void
    {
        try {
            $this->logger->info(sprintf(
                'Starting scenario(s): %s %s',
                implode(', ', $scenarios),
                $this->packageManager->getPrettyInfo()
            ));

            $this->mergedScenarios = $this->merger->merge($scenarios);
            $steps = $this->mergedScenarios['steps'];

            array_walk($steps, function (StepInterface $step, string $name) {
                $this->logger->debug('Running step: ' . $name);

                $step->execute();

                $this->logger->debug(sprintf('Step "%s" finished', $name));
            });

            $this->logger->info('Scenario(s) finished');
        } catch (StepException $stepException) {
            try {
                $actions = $this->mergedScenarios['actions'];

                array_walk($actions, function (ActionInterface $action, string $name) {
                    $this->logger->debug('Running on fail action: ' . $name);

                    $action->execute();

                    $this->logger->debug(sprintf('On fail action "%s" finished', $name));
                });
            } catch (ActionException $actionException) {
                $this->logger->error($actionException->getMessage());
            }
            $this->handleException($stepException);
        } catch (Throwable $exception) {
            $this->handleException(
                $exception,
                sprintf('Unhandled error: [%d] %s', $exception->getCode(), $exception->getMessage())
            );
        }
    }
```

Comment (NadiyaS) [reply src/Scenario/Processor.php:86]: no, we should catch Throwable, otherwise there may be a case when we did not set flag that deploy was failed and post-deploy phase will be executed.

**Review by oshmyheliuk — APPROVED** [review]

**Review by BaDos — APPROVED** [review]

**Comment by andriyShevtsov** [comment]
QA approved

File: src/Step/Deploy/SplitDbConnection.php (hunk 5 above)
Comment (andriyShevtsov) [reply src/Step/Deploy/SplitDbConnection.php]: @BarnyShergold please approve. Requested changes were implemented

**Review by BarnyShergold — APPROVED** [review]
